### ISSUE TITLE
Support synonyms

### ISSUE THREAD

**Issue by Anaphory** [lmaurits/BEASTling#144]
Binary models can, and therefore should, support multiple different values (eg. synonyms, for lexical data) in the same feature-/lect slot. Possibly not in beastling-style data files, but CLDFs can already represent it.

I hope this does not require too much modification of the internals.

**Comment by LinguList** [comment]
yes, this is essential. Does it have to do with the missing ID? As of now, CLDF usually has the ID field in the first column, but it's not needed for current beastling.

**Comment by Anaphory** [comment]
No, this doesn't really have to do with the ID column, it's more a thing of how Beastling represents values internally, which currently does not permit multiple values to be associated to a (*Language*, *Feature*) pair.

Changing this does require some architectural changes and also a decision – or a way to defer that decision to the user – on how to deal with this kind of value in non-binary models.

**Comment by LinguList** [comment]
well, the ID should be supported anyway, as this should ideally not be implicitly inferred by counting row numbers. So it should not be a language/feature pair, but an ID, with features and languages being attriburtes of that ID>.

**Comment by Anaphory** [comment]
That's an entirely different issue, which comes from us apparently never having caught up with the released CLDF guidelines. I'll raise that as #147.

**Comment by lmaurits** [comment]
Yeah, I keep hearing this from linguists. :(

I don't think that there is any sensible way to handle this kind of data with non-binary models, so if BEASTling detects this kind of data in a dataset that's not modelled with Covarion I would be in favour of it complaining and dying.  Thankfully this is not a big problem as it's not (yet!) at all common to use multistate data for cognates, and I don't _think_ this issue appears in structural data (famous last words).

If I'm not misremembering, whether you use CLDF or BEASTling's "house format", when all is said and done the data points end up represented as strings in a big 2D dictionary, e.g. `data[lang][feature] = "value"`.  The actual value string are arbitrary, and when serialised to XML they get replaced by their indices in an alphabetically sorted list of all possible strings (I'd really prefer it if the arbitrary strings ended up in the BEAST XML, but in the early days I didn't know enough about BEAST DataTypes to make this work reliably, and I think it would just result in larger XML files anyway).  I suppose the logical way to include synonyms would be to use a tuple of strings in place of a string.

This might not be as hard as I'd always assumed it would be...

Once we are using pycldf to read the latest and greatest CLDF format (and hence there is some way to actually read this kind of data into BEASTling) I will take a crack at this.

**Comment by Anaphory** [comment]
> not (yet!)

My feeling is more “not (any more!)”, because naïve multi-state models had even more problems than many of binarized models I see right now. We might come back to non-naïve multi-state models, but then hopefully from a linguistics perspective with some awareness of hurdles like synonyms.

Something like this *might* pop up in structural data later, eg. when looking at “Order of Subject, Verb and Object” having a value of “It can be SVO or SOV, but it's never anything else”.

Your description of what's going on internally matches what I remember and read in `fileio/datareaders.py`. Let's indeed deal with CLDF first, and then let's carefully think about what certain ways of implementing synonyms and similar ambiguities means, juggling a few reasonable test cases and their edgy generalizations (eg. empty tuple vs. no-value-specified-means-unknown, for your suggestion above; I can also imagine some kinds of probabilities or other scores associated with values, muse about how this might need to look like in native-beastling-csv or if at all, etc.) before committing on one.

**Comment by lmaurits** [comment]
Yeah, this one definitely requires some careful thought...

BEASTling hackathon at the centroid of Leiden, Jena and Turku? :P :beers:

**Comment by xrotwang** [comment]
Yes, that's one thing I hope for CLDF: that it becomes the format to define
binarization on top.

Am 13.09.2017 2:48 PM schrieb "Luke Maurits" <notifications@github.com>:

**Comment by lmaurits** [comment]
In the sense that people will start programmatically deriving binarised data from CLDF files instead of sharing binary data around as if it were a first class citizen?

**Comment by xrotwang** [comment]
exactly

**Comment by Anaphory** [comment]
Now that there is CLDF support for various cognate classifications, and BEASTling support for CLDF, we should take this back on the agenda.

**Comment by LinguList** [comment]
we have a new function for binarised output (probably the most consistent that is out there) in lingpy2.6: https://github.com/lingpy/lingpy/blob/master/lingpy/convert/strings.py#L404 

it is thought to produce nexus files, but it is probably useful to think of the various aspects, especially since binarisation creates problems with missing data if you take concepts into account (not if you ignore them).

**Comment by Anaphory** [comment]
I have looked at this today. Currently, the implicit behaviour of Beastling when presented with synonyms is to take the last occuring cognate class for that language/concept pair in a file, because it overrides. Other options (which could be selected by a switch defaulting to `last` – though by now it pains me to add more config options) would be to take the first occuring value or the set of all occuring values or the tuple of all occuring values in the file. That is, the file

    ID,Language_ID,Feature_ID,Value
    1,l1,c,1
    2,l1,c,2
    3,l2,c,1
    4,l3,c,2
    5,l4,c,2
    6,l4,c,1

could result in the following values for `c`, in order of how useful I think they are:

    l1:  2, l2: 1, l3: 2, l4:  1 #last
    l1: 12, l2: 1, l3: 2, l4: 12 #set
    l1:  1, l2: 1, l3: 2, l4:  2 #first
    l1: 12, l2: 1, l3: 2, l4: 21 #order

I *think* I have undestood how handling synonyms might work in practice. One minor issue is that it requires that `BaseModel`'s `compute_feature_properties(self)` would have to be overloaded by `BinaryModel` doing very much the same, maybe I find a way to unify that behaviour.

**Comment by Anaphory** [comment]
I seem to have built something that runs in some cases already. This is a note to myself that we some tests for non-obvious constant features (simple: all languages either `?` or `V`; complicated: All languages either `?` or `V, W` – the first should easily show up as constant feature because `V` is the only value. The second one would be constant in each sub-feature, but differentiating it from the case where even a single language has `V` instead of `V, W` will not be easy.)

**Comment by lmaurits** [comment]
We do support synonyms now, right @Anaphory?  Do we still need this Issue?  If the synonym support has some small details which need improvement maybe they should become separate Issues?

**Comment by Anaphory** [comment]
We should have synonym support, yes. Let me see whether it's in the tests to confirm.
